src/Compilers/CSharp/Portable/Symbols/ArrayTypeSymbol.cs

```diff
@@ -404,9 +404,44 @@ internal override bool ApplyNullableTransforms(byte defaultTransformFlag, Immuta
             return true;
         }
 
-        internal override TypeSymbol SetNullabilityForReferenceTypes(Func<TypeWithAnnotations, TypeWithAnnotations> transform)
+        internal override TypeSymbol SetObliviousNullabilityForReferenceTypes()
         {
-            return WithElementType(transform(ElementTypeWithAnnotations));
+            if (ElementType.TypeKind == TypeKind.Array)
+            {
+                // The language supports deeply nested arrays, up to 10,000+ instances. This means we can't implemented a head
+                // recursive solution here. Need to take an iterative approach to building up the annotations here.
+                var builder = ArrayBuilder<ArrayTypeSymbol>.GetInstance();
+
```

Comment:
This is good, but a stack would be clearer, if you were motivated to add a pooled stack.

Reply:
Good point. See so many `ArrayBuilder` I sometimes forget that we have stacks 😉

Reply:
Consider using `ArrayBuilderExtensions.Push()` and `Pop()` instead.

---
In reply to: [273743425](https://github.com/dotnet/roslyn/pull/34891#discussion_r273743425) [](ancestors = 273743425)

Reply:
Agreed. Would prefer a stack here, it would make the second part of this much easier to reason about.